File: proofchecker/syntax.py
    # TODO: In find_main_operator(), test for multiple operators at depth 0 
    # Apply standard order of operations if multiple operators at depth 0
    # e.g. ¬A∨B should recognize ∨ as the main logical operator

class Syntax:

    @staticmethod
    def is_valid_TFL(str):
        """
        Determine if a line of text represents a valid TFL statement
        """
        
        # Lines in a `prooftext` contain a justification after the '#' symbol
        # If the line contains a justification, remove it
        line = str
        if '#' in line:
            line = Syntax.remove_justification(line)

        # Strip all whitespace from the line
        line.replace(" ", "")

        # Verify all remaining characters are valid TFL symbols
        if Syntax.has_valid_symbols(line):

            # Check for matching parentheses
            if Syntax.has_balanced_parens(line):

                # Find the main logical operator
                if ('∧' or '∨' or '¬' or '→' or '↔') in line:
                    opIndex = Syntax.find_main_operator(line)
                    
                    # Grab the substrings around the main operator
                    left = line[0:opIndex]
                    right = line[opIndex+1:len(line)]

                    # There should be values on both sides unless the operator is ¬
                    if not (line[opIndex] == '¬'):
                        if left == '':
                            return False
                        if right == '':
                            return False

                    # Determine that both substrings are valid TFL sentences (recursion)
                    leftIsValid = Syntax.is_valid_TFL(left)
                    rightIsValid = Syntax.is_valid_TFL(right)
                    
                    if not (leftIsValid and rightIsValid):
                        return False

            else:
                return False
        
        else:
            return False

        # If we reached this line, everything checks out
        return True



    @staticmethod
    def remove_justification(str):
        """
        Removes the justification from the line, if present
        """

        line = str
        char = '#'
        if char in line:
            x = 0
            for char in line:
                if char == '#':
                    line = line[0:x]
                    break
                else:
                    x += 1

        return line

    @staticmethod
    def has_balanced_parens(str):
        """
        Determines if a string has balanced parentheses
        """
        OPEN_PARENS=['(', '[', '{']
        CLOSED_PARENS=[')', ']', '}']
        stack = []

        for char in str:
            if char in OPEN_PARENS:
                stack.append(char)
            elif char in CLOSED_PARENS:
                pos = CLOSED_PARENS.index(char)
                if ((len(stack) > 0) and 
                    (OPEN_PARENS[pos] == stack[len(stack)-1])):
                    stack.pop()
                else:
                    return False
        if len(stack)==0:
            return True
        else:
            return False

            
    @staticmethod
    def has_valid_symbols(str):
        """
        Verifies that all characters in a string are valid TFL symbols
        """
        # TODO: Replace ATOMIC with a more elegant regular expression
        ATOMIC = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I',
            'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S',
            'T', 'U', 'V', 'W', 'X', 'Y', 'Z']
        CONNECTIVES = ['∧', '∨', '¬', '→', '↔']
        PARENS = ['(', '[', '{', ')', ']', '}']

        for char in str:
            if not ((char in ATOMIC) or (char in CONNECTIVES) or (char in PARENS)):
                return False

        return True

    @staticmethod
    def find_main_operator(str):
        """
        Returns the index of the main logical operator in a TFL sentence
        """
        CONNECTIVES = ['∧', '∨', '¬', '→', '↔']
        
        line = str
        opIndex = 0

        # Determine the depth of each char in the sentence
        depthArray = Syntax.set_depth_array(line)

        # Remove matching outermost parentheses
        if depthArray[0]==1:
            parensMatch = True
            index = 0
            # Depth drops to zero somewhere if outermost parentheses do not match
            for char in line[0:len(line)-1]:
                parensMatch = (parensMatch and depthArray[index]>0)
                index += 1
            # Strip the outermost parentheses, call function recursively
            # Add one to the result for the leading parenthesis that was removed
            if parensMatch:
                return (Syntax.find_main_operator(line[1:len(line)-1]) + 1)

        # Find the main operator
        for char in line:
            if ((char in CONNECTIVES) and (depthArray[opIndex]==0)):
                return opIndex
            else:
                opIndex += 1

        # If no operator found, return 0
        return 0

    @staticmethod
    def set_depth_array(str):
        """
        Returns an array containing the depth of each character in a TFL sentence
        """
        OPEN_PARENS=['(', '[', '{']
        CLOSED_PARENS=[')', ']', '}']

        depth = 0
        depthArray = []

        for char in str:
            if char in OPEN_PARENS:
                depth += 1
            elif char in CLOSED_PARENS:
                depth -= 1
            
            depthArray.append(depth)
        
        return depthArray

File: proofchecker/static/proofchecker/js/syntax.js
/*
A Javascript equivalent to the current Syntax Python file that valid_input.js utilizes to check if the user's input is valid
*/

const ATOMIC = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I',
'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S',
'T', 'U', 'V', 'W', 'X', 'Y', 'Z'];

const CONNECTIVES = ['∧', '∨', '¬', '→', '→', '↔','^', '&', 'v', '>', '->', '<->', '-', '~']
const PARENS = ['(', '[', '{', ')', ']', '}'];
const OPEN_PARENS=['(', '[', '{'];
const CLOSED_PARENS=[')', ']', '}'];
const DELIMITERS = [',', ' '];

function isValidTFL(string) {

    /*
    Determine if a line of text represents a valid TFL statement
    Lines in a `prooftext` contain a justification after the '#' symbol
    If the line contains a justification, remove it
    */

    let line = "";
    line = string;
    if(line.includes('#'))
    {
        removeJustification(line);
    }
    
    //Strip all whitespace from the line
    line.replace(/\s+/g, '');

    //Verify all remaining characters are valid TFL symbols
    if(hasValidSymbols(line))
    {
        //Check for matching parentheses
        if(hasBalancedParens(line))
        {
            //Determine the depth of each char in the line
            let depthArray = setDepthArray(line);

            // Remove matching outermost parentheses
            if(depthArray[0] == 1)
            {
                let parenthesesMatch = true;
                let index = 0;

                //Depth drops to zero somewhere if outermost parentheses do not match
                for(var i = 0; i < line.length-1; i++)
                {
                    parenthesesMatch = parenthesesMatch && (depthArray[index] > 0);
                    index++;
                }

                //Strip the outermost parentheses, call function recursively
                //Add one to the result for the leading parenthesis that was removed
                if(parenthesesMatch)
                {
                    return isValidTFL(line.substring(1, line.length-1));
                }

                //Find the main logical operator
                if (line.includes('∧') || line.includes('∨') || line.includes('¬') || line.includes('→') || line.includes('↔'))
                {
                    let operatorIndex = findMainOperator(line);

                    //Grab the substrings around the main operator
                    let left = line.substring(0, operatorIndex);
                    let right = line.substring(operatorIndex+1, line.length);

                    //There should be values on both sides unless the operator is ¬
                    if(line.charAt(operatorIndex) != '¬')
                    {
                        if(left == '')
                        {
                            return "The left side of the main logical operator (" + line.charAt(operatorIndex) + ") is empty!";
                            //return false;
                        }
                        if(right == '')
                        {
                            return "The right side of the main logical operator (" + line.charAt(operatorIndex) + ") is empty!";
                            //return false;
                        }
                    }

                    //Determine that both substrings are valid TFL sentences (recursion)
                    let leftIsValid = isValidTFL(left);
                    let rightIsValid = isValidTFL(right);

                    if(!(leftIsValid && rightIsValid))
                    {                       
                        let substringError = "";

                        if(!leftIsValid)
                        {
                            substringError = substringError + "The left substring of the main operator (" + line.charAt(operatorIndex) + ") is invalid! ";
                        }
                               
                        if(!rightIsValid)
                        {
                            substringError = substringError + "The right substring of the main operator (" + line.charAt(operatorIndex) + ") is invalid!";
                        }
                     
                        
                         
                        return substringError;
                    }
                }
            }    
        }            
        else{
            return "This statement does not have balanced parantheses." ;
        }
    }
    else{
        return "This statement has at least one invalid TFL symbol." ;
    }

    //If we reached this line, everything checks out

    return "This is a valid TFL statement.";

}

function removeJustification(string){
    //Removes the justification from the line, if present
    let line = "";
    line = string;

    let char = '#';
    if(line.includes(char))
    {
        let endIndex = line.indexOf(char);
        line = line.substring(0, endIndex);
    }

    return line;

}

function hasBalancedParens(string){
    //Determines if a string has balanced parentheses
    let stack = [];
    let line = "";
    line = string;

    for(var i = 0; i < line.length; i++)
    {
        if(OPEN_PARENS.includes(line.charAt(i)))
        {
            stack.push(line.charAt(i));
        }
        else if(CLOSED_PARENS.includes(line.charAt(i)))
        {
            let position = CLOSED_PARENS.indexOf(line.charAt(i));
            if(stack.length > 0 && OPEN_PARENS[position] == stack[stack.length - 1])
            {
                stack.pop();
            }
            else{
                return false;
            }
        }
    }

    if(stack.length == 0)
    {
        return true;
    }

    return false;

}

function hasValidSymbols(string){

    let line = "";
    line = string;

    //Verifies that all characters in a string are valid TFL symbols
    for(var i = 0; i < line.length; i++)
    {
        if(!(ATOMIC.includes(line.charAt(i)) || CONNECTIVES.includes(line.charAt(i)) || PARENS.includes(line.charAt(i)) || DELIMITERS.includes(line.charAt(i))))
        {
            return false;
        }
    }

    return true;

}

function findMainOperator(string){

    //Returns the index of the main logical operator in a TFL sentence
    let line = "";
    line = string;

    let operatorIndex = 0;

    //Determine the depth of each char in the sentence
    let depthArray = setDepthArray(line);

    //Remove matching outermost parentheses
    if(depthArray[0] == 1)
    {
        let parenthesesMatch = true;
        let index = 0;

        //Depth drops to zero somewhere if outermost parentheses do not match
        for(var i = 0; i < line.length-1; i++)
        {
            parenthesesMatch = (parenthesesMatch && depthArray[index] > 0);
            index++;
        }

        //Strip the outermost parentheses, call function recursively
        //Add one to the result for the leading parenthesis that was removed
        if(parenthesesMatch)
        {
            line = line.substring(1, line.length-1);
            return findMainOperator(line) + 1;
        }

    }

    //Find the main operator
    for(var i = 0; i < line.length; i++)
    {
        if(CONNECTIVES.includes(line.charAt(i)) && depthArray[operatorIndex] == 0)
        {
            return operatorIndex;
        }
        else{
            operatorIndex++;
        }

    }

    //If no operator found, return 0
    return 0;        


}

function setDepthArray(string){

    //Returns an array containing the depth of each character in a TFL sentence

    let depth = 0;
    let depthArray = [];

    let line = "";
    line = string;

    for(var i = 0; i < line.length; i++)
    {
        if(OPEN_PARENS.includes(line.charAt(i)))
        {
            depth++;
        }
        else if(CLOSED_PARENS.includes(line.charAt(i)))
        {
            depth--;
        }

        depthArray.push(depth);
    }

    return depthArray;
}